pkg/apiserver/debugapi/endpoint.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debugapi
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+
+	"github.com/pingcap/tidb-dashboard/pkg/apiserver/model"
+)
+
+var (
+	ErrPathParamNotMatched = ErrNS.NewType("path_param_not_matched")
+	ErrValueTransformed    = ErrNS.NewType("error_value_transformed")
+)
+
+type ValueGetter interface {
+	Get(key string) string
+}
+
+type EndpointAPI struct {
+	ID        string             `json:"id"`
+	Component model.NodeKind     `json:"component"`
+	Path      string             `json:"path"`
+	Method    string             `json:"method"`
+	Host      EndpointAPIParam   `json:"host"`
+	Segment   []EndpointAPIParam `json:"segment"` // e.g. /stats/dump/{db}/{table} -> db, table
+	Query     []EndpointAPIParam `json:"query"`   // e.g. /debug/pprof?seconds=1 -> seconds
+}
+
+func (e *EndpointAPI) NewRequest(value ValueGetter) (*http.Request, error) {
+	host, err := e.PopulateHost(value)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "populate host")
+	}
+
+	r, err := http.NewRequest(e.Method, fmt.Sprintf("//%s", host), nil)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "new request")
+	}
+
+	// we can put param and value together at first
+	pValues := newParamValues(e.Segment, value)
+	path, err := e.PopulatePath(pValues)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "populate path")
+	}
+	r.URL.Path = path
+
+	qValues := newParamValues(e.Query, value)
+	rawQuery, err := e.EncodeQuery(qValues)
+	if err != nil {
+		return nil, ErrValueTransformed.Wrap(err, "encode queries")
+	}
+	r.URL.RawQuery = rawQuery
+
+	return r, nil
+}
+
+func (e *EndpointAPI) PopulateHost(value ValueGetter) (string, error) {
+	return e.Host.Transform(value.Get(e.Host.Name))
+}
+
+var ParamRegexp *regexp.Regexp = regexp.MustCompile(`\{(\w+)\}`)
+
+func (e *EndpointAPI) PopulatePath(pvs paramValues) (string, error) {
+	var returnErr error
+	replacedPath := e.Path
+	replacedPath = ParamRegexp.ReplaceAllStringFunc(replacedPath, func(s string) string {
+		if returnErr != nil {
+			return s
+		}
+
+		key := ParamRegexp.ReplaceAllString(s, "${1}")
+		pv, ok := pvs.Find(key)
+		// means the param can be found in the endpoint path, but not in the endpoint param
+		if !ok {
+			returnErr = ErrPathParamNotMatched.New("path: %s, param: %s", e.Path, key)
```

Comment:
According to your comment in L89, this error is more like, missing required parameter?